### PR TITLE
fix allowed origins to make webui work again

### PR DESCRIPTION
**Pull request by whyrusleeping**
I'm not really sure what these changes mean, All I know is that the webui now works and I can add files through that interface again.

cc @mappum @diasdavid @jbenet and anyone else who can tell me if what i've done here is right.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

**Comment by whyrusleeping** [comment]
I would also love to have some tests to somehow make sure the webui works against each ipfs commit so we dont accidentaly break things as we move along.

File: commands/http/handler.go [hunk 1]
```diff
@@ -56,10 +57,8 @@ const (
 )
 
 var localhostOrigins = []string{
-	"http://127.0.0.1",
-	"https://127.0.0.1",
-	"http://localhost",
-	"https://localhost",
+	"127.0.0.1:5001",
+	"localhost:5001",
 }
```
Comment (mappum) [inline commands/http/handler.go]: We should probably get the API port from the config rather than hardcoding port 5001.

**Comment by mappum** [comment]
Other than my inline comment, this looks good. It's important to restrict access only to the API server, instead of all localhost origins/referers, otherwise attacks can still come from pages served over the local gateway (http://localhost:8080/ipfs/EVIL_PAGE).

File: commands/http/handler.go [hunk 2]
```diff
@@ -347,12 +346,17 @@ func allowOrigin(r *http.Request, cfg *ServerConfig) bool {
...
+	if err != nil {
+		log.Error("incorrectly formatted origin on request: ", origin)
+	}
+
 	for _, o := range cfg.CORSOpts.AllowedOrigins {
 		if o == "*" { // ok! you asked for it!
 			return true
 		}
 
-		if o == origin { // allowed explicitly
```
Comment (jbenet) [inline commands/http/handler.go]: no, i think the protocol matters to CORS, one should be able to set CORS on https only :-1:

File: commands/http/handler.go (hunk 1 above)
Comment (jbenet) [inline commands/http/handler.go]: :+1: to @mappum's comment. should try to grab it from config.

i think this is the key part: adding the API port, not removing the protocol. this should work then:

```go
var localhostOrigins = []string{
 "http://127.0.0.1:<api-port>", 
 "https://127.0.0.1:<api-port>",
 "http://localhost:<api-port>",   
 "https://localhost:<api-port>",    
 }
```

File: commands/http/handler.go [hunk 3]
```diff
@@ -384,7 +394,7 @@ func allowReferer(r *http.Request, cfg *ServerConfig) bool {
 		}
 
 		// referer is allowed explicitly
-		if o == referer {
```
Comment (jbenet) [inline commands/http/handler.go]: same as above

File: commands/http/handler.go (hunk 1 above)
Comment (whyrusleeping) [inline commands/http/handler.go]: but the protocol isnt part of the url host

File: core/corehttp/commands.go [hunk 4]
```diff
@@ -69,6 +76,16 @@ func CommandsOption(cctx commands.Context) ServeOption {
...
 
+		rcfg := n.Repo.Config()
+		apiaddr := strings.Split(rcfg.Addresses.API, "/")
+		port := apiaddr[len(apiaddr)-1]
+
+		var origins []string
+		for _, o := range localhostOrigins {
+			origins = append(origins, o+":"+port)
+		}
+		cfg.CORSOpts.AllowedOrigins = origins
```
Comment (jbenet) [inline core/corehttp/commands.go]: incorect. overwriting `AllowedOrigins` always

File: core/corehttp/commands.go [hunk 5]
```diff
@@ -57,6 +57,13 @@ func addHeadersFromConfig(c *cmdsHttp.ServerConfig, nc *config.Config) {
 	c.Headers = nc.API.HTTPHeaders
 }
 
+var localhostOrigins = []string{
+	"https://localhost",
+	"http://localhost",
+	"https://127.0.0.1",
+	"http://127.0.0.1",
+}
```
Comment (jbenet) [inline core/corehttp/commands.go]: incorrect. these constants are already in `commands/http/handler`

**Comment by jbenet** [comment]
thanks for helping fix this @whyrusleeping :+1:

**Comment by whyrusleeping** [comment]
@jbenet thanks for actually fixing it :smile:

**Comment by krl** [comment]
@whyrusleeping the api has a test suite that should probably be run on go-ipfs changes as well, i made the tests use browserify/phantomjs as well

**Comment by whyrusleeping** [comment]
@krl  that would be pretty sweet to have. How would we hook up those tests to the go-ipfs CI?

**Pull request merged**